In [2]:
# Update sklearn to prevent version mismatches
# !pip install sklearn --upgrade
# install joblib to save model. 
# !pip install joblib

In [14]:
import pandas as pd
import tensorflow
import numpy as np

In [15]:
df = pd.read_csv("../Resources/cleaned_data.csv")
df

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  \
0          CONFIRMED              0              0              0   
1          CONFIRMED              0              0              0   
2     FALSE POSITIVE              0              1              0   
3     FALSE POSITIVE              0              1              0   
4          CONFIRMED              0              0              0   
...              ...            ...            ...            ...   
8739  FALSE POSITIVE              0              0              0   
8740  FALSE POSITIVE              0              1              1   
8741       CANDIDATE              0              0              0   
8742  FALSE POSITIVE              0              0              1   
8743  FALSE POSITIVE              0              0              1   

      koi_fpflag_ec  koi_period  koi_period_err1  koi_period_err2  \
0                 0    9.488036     2.775000e-05    -2.775000e-05   
1                 0   54.418383     2.479000e-04    -2.479000e-04   
2                 0   19.899140     1.494000e-05    -1.494000e-05   
3                 0    1.736952     2.630000e-07    -2.630000e-07   
4                 0    2.525592     3.761000e-06    -3.761000e-06   
...             ...         ...              ...              ...   
8739              1    8.589871     1.846000e-04    -1.846000e-04   
8740              0    0.527699     1.160000e-07    -1.160000e-07   
8741              0    1.739849     1.780000e-05    -1.780000e-05   
8742              0    0.681402     2.434000e-06    -2.434000e-06   
8743              1    4.856035     6.356000e-05    -6.356000e-05   

      koi_time0bk  koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  \
0      170.538750          0.002160  ...           -81.0      4.467   
1      162.513840          0.003520  ...           -81.0      4.467   
2      175.850252          0.000581  ...          -176.0      4.544   
3      170.307565          0.000115  ...          -174.0      4.564   
4      171.595550          0.001130  ...          -211.0      4.438   
...           ...               ...  ...             ...        ...   
8739   132.016100          0.015700  ...          -152.0      4.296   
8740   131.705093          0.000170  ...          -166.0      4.529   
8741   133.001270          0.007690  ...          -220.0      4.444   
8742   132.181750          0.002850  ...          -236.0      4.447   
8743   135.993300          0.010800  ...          -225.0      4.385   

      koi_slogg_err1  koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2  \
0              0.064          -0.096     0.927          0.105         -0.061   
1              0.064          -0.096     0.927          0.105         -0.061   
2              0.044          -0.176     0.868          0.233         -0.078   
3              0.053          -0.168     0.791          0.201         -0.067   
4              0.070          -0.210     1.046          0.334         -0.133   
...              ...             ...       ...            ...            ...   
8739           0.231          -0.189     1.088          0.313         -0.228   
8740           0.035          -0.196     0.903          0.237         -0.079   
8741           0.056          -0.224     1.031          0.341         -0.114   
8742           0.056          -0.224     1.041          0.341         -0.114   
8743           0.054          -0.216     1.193          0.410         -0.137   

             ra        dec  koi_kepmag  
0     291.93423  48.141651      15.347  
1     291.93423  48.141651      15.347  
2     297.00482  48.134129      15.436  
3     285.53461  48.285210      15.597  
4     288.75488  48.226200      15.509  
...         ...        ...         ...  
8739  298.74921  46.973351      14.478  
8740  297.18875  47.093819      14.082  
8741  286.50937  47.163219      14.757  
8742  294.16489  47.176281      15.385  
8743  297.00977  47.121021      14.826  

[8744 rows x 41 columns]

# Create a Train Test Split

Use `koi_disposition` for the y values

In [16]:
from sklearn.model_selection import train_test_split
y = df['koi_disposition']
X = df.drop(columns='koi_disposition')

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 1, stratify = y)

In [17]:
X_train.head()

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
5496              1              0              0              0  252.044440   
8595              1              0              0              0  371.518520   
3963              0              0              1              0    8.038670   
5467              0              0              0              0   18.782160   
6376              1              0              1              0   12.429716   

      koi_period_err1  koi_period_err2  koi_time0bk  koi_time0bk_err1  \
5496         0.027490        -0.027490     265.2010            0.0494   
8595         0.015790        -0.015790     317.6836            0.0339   
3963         0.000114        -0.000114     135.3098            0.0123   
5467         0.000406        -0.000406     147.8508            0.0148   
6376         0.000472        -0.000472     141.2846            0.0420   

      koi_time0bk_err2  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
5496           -0.0494  ...          -136.0      4.621           0.041   
8595           -0.0339  ...          -206.0      4.377           0.101   
3963           -0.0123  ...          -181.0      4.485           0.050   
5467           -0.0148  ...          -167.0      4.488           0.048   
6376           -0.0420  ...          -200.0      4.534           0.037   

      koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
5496          -0.035     0.664          0.057         -0.059  292.79022   
8595          -0.203     1.089          0.364         -0.145  293.06400   
3963          -0.200     0.975          0.282         -0.101  290.51785   
5467          -0.290     0.940          0.386         -0.087  291.76413   
6376          -0.213     0.905          0.281         -0.088  297.52072   

            dec  koi_kepmag  
5496  41.948639      15.884  
8595  45.034210      13.731  
3963  41.238762      14.999  
5467  41.860130      14.043  
6376  40.585419      15.842  

[5 rows x 40 columns]

# Train the Model

In [20]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
model = LogisticRegression(max_iter = 1000)

In [21]:
model.fit(X_train, y_train)
print(f"The cross-validated accuracy for unprocessed training set:\n{round(cross_val_score(model,X_train, y_train,scoring='accuracy', cv=5).mean(),2)}")
# failed to converge without data's being scaled or standardized: accuracy: 66%

/Users/jiaping/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Users/jiaping/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#l

The cross-validated accuracy for unprocessed training set:
0.66


/Users/jiaping/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [22]:
print(f"The cross-validated accuracy for normalized training set: \n{round(cross_val_score(model,X_train_scaled, y_train,scoring='accuracy', cv=5).mean()*100,2)}%")
print(f"The cross-validated accuracy for standardized training set:\n{round(cross_val_score(model,X_train_stand, y_train,scoring='accuracy', cv=5).mean()*100,2)}%")

The cross-validated accuracy for normalized training set: 
84.89%
The cross-validated accuracy for standardized training set:
88.67%


Standardization produces higher cross-validated accuracy than normalization, so standardized data will be used for hyperparameter tuning.

# Hyperparameter Tuning

Use `GridSearchCV` to tune the model's parameters

In [23]:
from sklearn.model_selection import GridSearchCV,RandomizedSearchCV
from sklearn.metrics import classification_report

param_grid = {'C': [0.00001, 0.0001,0.001,0.01,0.1, 1, 10, 100, 1000], 
              'max_iter': [1000, 5000, 10000],
              'penalty': ["l1", "l2"],
              'solver': ['liblinear', 'lbfgs']}

# Because the "liblinear" solver does not improve the model performance and also makes the search much slower compared to "lbfgs" solver
# grid search about solver and penalty is deleted (lbfgs solver only deal with l2 penalty) for the gridsearch on standardized data
param_grid2 = {'C': [0.00001, 0.0001,0.001,0.01,0.1, 1, 10, 100, 1000], 
              'max_iter': [1000, 5000, 10000]}
model = LogisticRegression()


### 1. Normalized data

In [24]:
grid1 = GridSearchCV(model, param_grid, verbose = 3, n_jobs = -1)
grid1.fit(X_train_scaled, y_train)

Fitting 5 folds for each of 108 candidates, totalling 540 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  24 tasks      | elapsed:    2.2s
[Parallel(n_jobs=-1)]: Done 376 tasks      | elapsed:   25.7s
[Parallel(n_jobs=-1)]: Done 540 out of 540 | elapsed:  4.8min finished
/Users/jiaping/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


GridSearchCV(estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': [1e-05, 0.0001, 0.001, 0.01, 0.1, 1, 10, 100,
                               1000],
                         'max_iter': [1000, 5000, 10000],
                         'penalty': ['l1', 'l2'],
                         'solver': ['liblinear', 'lbfgs']},
             verbose=3)

In [25]:
print(f"The best paramaters for the tuned model: {grid1.best_params_}\n") 
print(f"The cross-validated accuracy for the tuned model with normalized training set:{round(grid1.best_score_*100,2)}%")

The best paramaters for the tuned model: {'C': 1000, 'max_iter': 1000, 'penalty': 'l2', 'solver': 'lbfgs'}

The cross-validated accuracy for the tuned model with normalized training set:88.9%


After model tuning, the cross-validated accuracy is increased from 84.89% to 88.9%.

Results on the whole training set

In [26]:
predictions1 = grid1.predict(X_train_scaled)
print(classification_report(y_train, predictions1))
print(f"{round(grid1.score(X_train_scaled, y_train)*100,2)}%")  # 88.44%, 89.3%

                precision    recall  f1-score   support

     CANDIDATE       0.84      0.70      0.76      1586
     CONFIRMED       0.76      0.86      0.81      1704
FALSE POSITIVE       0.98      1.00      0.99      3268

      accuracy                           0.89      6558
     macro avg       0.86      0.85      0.85      6558
  weighted avg       0.89      0.89      0.89      6558

88.9%


### 2. Standardized data

In [27]:
grid2 = GridSearchCV(model, param_grid2, verbose = 3, n_jobs = -1)
grid2.fit(X_train_stand, y_train)

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.


Fitting 5 folds for each of 27 candidates, totalling 135 fits


[Parallel(n_jobs=-1)]: Done  40 tasks      | elapsed:    0.9s
[Parallel(n_jobs=-1)]: Done 135 out of 135 | elapsed:   33.1s finished


GridSearchCV(estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': [1e-05, 0.0001, 0.001, 0.01, 0.1, 1, 10, 100,
                               1000],
                         'max_iter': [1000, 5000, 10000]},
             verbose=3)

In [28]:
print(f"The best paramaters for the tuned model: {grid2.best_params_}\n")
print(f"The cross-validated accuracy for the tuned model with standardized training set:{round(grid2.best_score_*100,2)}%")

The best paramaters for the tuned model: {'C': 1000, 'max_iter': 5000}

The cross-validated accuracy for the tuned model with standardized training set:89.04%


After model tuning, the cross-validated accuracy is increased from 88.67% to 89.04%.

Results on the whole training set

In [29]:
predictions2 = grid2.predict(X_train_stand)
print(classification_report(y_train, predictions2))
print(f"{round(grid2.score(X_train_stand, y_train)*100,2)}%")

                precision    recall  f1-score   support

     CANDIDATE       0.84      0.71      0.77      1586
     CONFIRMED       0.77      0.86      0.82      1704
FALSE POSITIVE       0.98      1.00      0.99      3268

      accuracy                           0.89      6558
     macro avg       0.87      0.86      0.86      6558
  weighted avg       0.89      0.89      0.89      6558

89.34%


The tuned model with standardized data perfoems slightly better than the tuned model with normalized data, as the cross-validated accuracy of the tuned model with standardized data (89.04%) is higher than the one with normalized data (88.9%).

The final logistic regression model :
1. Requires the standardization of data.
2. Model parameters:  {'C': 10, 'penalty': 'l2', 'solver': 'lbfgs'}
3. Cross-validated accuracy: 89.04%

# Save the tuned model

In [33]:
import joblib
filename = 'logistic_regression.sav'
joblib.dump(grid2, filename)

['logistic_regression.sav']